protocol/app/ante/market_update_test.go

```diff
@@ -0,0 +1,554 @@
+					Markets:   nil,
+				},
+			},
+			want:    true,
+			wantErr: false,
+		},
+		{
+			name: "return error for multiple msg update",
+			msgs: []sdk.Msg{
+				&mmtypes.MsgUpdateMarkets{
+					Authority:     constants.BobAccAddress.String(),
+					UpdateMarkets: nil,
+				},
+				&mmtypes.MsgUpdateMarkets{
+					Authority:     constants.BobAccAddress.String(),
+					UpdateMarkets: nil,
+				},
+			},
+			want:    false,
+			wantErr: true,
+		},
+		{
+			name: "return error for multiple msg upsert",
+			msgs: []sdk.Msg{
+				&mmtypes.MsgUpsertMarkets{
+					Authority: constants.BobAccAddress.String(),
+					Markets:   nil,
+				},
+				&mmtypes.MsgUpsertMarkets{
+					Authority: constants.BobAccAddress.String(),
+					Markets:   nil,
+				},
+			},
+			want:    false,
+			wantErr: true,
+		},
+		{
+			name: "return error for multiple mixed market map messages",
+			msgs: []sdk.Msg{
+				&mmtypes.MsgUpsertMarkets{
+					Authority: constants.BobAccAddress.String(),
+					Markets:   nil,
+				},
+				&mmtypes.MsgUpdateMarkets{
+					Authority:     constants.BobAccAddress.String(),
+					UpdateMarkets: nil,
+				},
+			},
+			want:    false,
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			suite := testante.SetupTestSuite(t, true)
+			suite.TxBuilder = suite.ClientCtx.TxConfig.NewTxBuilder()
+
+			require.NoError(t, suite.TxBuilder.SetMsgs(tt.msgs...))
+
+			// Empty private key, so tx's signature should be  .
+			var (
+				privs   []cryptotypes.PrivKey
+				accSeqs []uint64
+				accNums []uint64
+			)
+
+			tx, err := suite.CreateTestTx(
+				suite.Ctx,
+				privs,
+				accNums,
+				accSeqs,
+				suite.Ctx.ChainID(),
+				signing.SignMode_SIGN_MODE_DIRECT,
+			)
+			require.NoError(t, err)
+
+			got, err := ante.IsMarketUpdateTx(tx)
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+
+			if got != tt.want {
+				t.Errorf("IsMarketUpdateTx() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+var (
+	testMarket = mmtypes.Market{
+		Ticker: mmtypes.Ticker{
+			CurrencyPair: types.CurrencyPair{
+				Base:  "TEST",
+				Quote: "USD",
+			},
+			Decimals:         1,
+			MinProviderCount: 1,
+			Enabled:          true,
+			Metadata_JSON:    "",
+		},
+		ProviderConfigs: nil,
+	}
+
+	testMarketParams = prices_types.MarketParam{
+		Id:                 0,
+		Pair:               "TEST-USD",
+		Exponent:           -8,
+		MinExchanges:       1,
+		MinPriceChangePpm:  10,
+		ExchangeConfigJson: `{"test_config_placeholder":{}}`,
+	}
+)
+
+func TestValidateMarketUpdateDecorator_AnteHandle(t *testing.T) {
+	type marketPerpPair struct {
+		market prices_types.MarketParam
+		perp   perpetualtypes.Perpetual
+	}
+
+	type args struct {
+		msgs        []sdk.Msg
+		simulate    bool
+		marketPerps []marketPerpPair
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "do nothing for non-market map messages - simulate",
```

Comment:
what's the purpose of testing the simulate flag?

Reply:
certain code paths are taken based on whether the antehandler is run in simulate or not.  as things stand with this specific antehandler, there arent differences based on simulation or not.  

But this gives us exhaustive test cases in case some simulation-based logic as added at any point.